src/sequential/sequential.hpp

```diff
@@ -0,0 +1,198 @@
+/// HELPER CLASSES
+////////////////////////////////////////////////////////////////////////////////
+
+enum SearchStatus {
+  FOUND_MATCH,
+  FOUND_SWAP,
+  FOUND_HOLE,
+  FOUND_NOHOLE,
+};
+
+/// @brief  Bucket for the Robin Hood hash table.
+struct SequentialBucket {
+  KeyType key;
+  ValueType value;
+  HashCodeType hashcode;
+  // A value of SIZE_MAX means that the bucket is empty. I do this hack so that
+  // we can fit this bucket into 4 words, which is more amenable to the hardware.
+  // A value of SIZE_MAX would be attrocious for performance anyways.
+  size_t offset;
+
+  SequentialBucket(KeyType key, ValueType value, HashCodeType hashcode, size_t offset) {
+    LOG_TRACE("Enter");
+    this->key = key;
+    this->value = value;
+    this->hashcode = hashcode;
+    this->offset = offset;
+  }
+
+  SequentialBucket() {
+    LOG_TRACE("Enter");
+    this->key = 0;
+    this->value = 0;
+    this->hashcode = 0;
+    this->offset = SIZE_MAX;
+  }
+
+  bool is_empty();
+
+  void set(const KeyType key,
+           const ValueType value,
+           const HashCodeType hashcode,
+           const size_t offset);
+
+  void swap(SequentialBucket &other);
+
+  bool equal_key(const KeyType key, const HashCodeType hashcode);
+
+  /// @brief  Pretty print bucket
+  ///
+  /// This prints in the format (using Python's f-string syntax):
+  /// f"({hashcode}=>{home}+{offset}) {key}: {value}"
+  void print(const size_t size);
+};
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// STATIC HELPER FUNCTIONS
+////////////////////////////////////////////////////////////////////////////////
+
+/// Source: https://stackoverflow.com/questions/664014/what-integer-hash-function-are-good-that-accepts-an-integer-hash-key
+HashCodeType
+hash(const KeyType key);
+
+size_t
+get_home(const HashCodeType hashcode, const size_t size);
+
+/// @brief  Get real bucket index.
+size_t
+get_probe(const size_t home, const size_t offset, const size_t size);
+
+/// @brief  Get offset from home or where it would be if not found.
+///         Return SIZE_MAX if no hole is found.
+std::pair<SearchStatus, size_t>
+get_wouldbe_offset(      std::vector<SequentialBucket> &tmp_buckets,
+                   const KeyType key,
+                   const HashCodeType hashcode,
+                   const size_t home);
+
+/// @brief  Insert but assume no resize is necessary.
+int
+insert_without_resize(      std::vector<SequentialBucket> &tmp_buckets,
+                          const KeyType key,
+                          const ValueType value,
+                          const HashCodeType hashcode);
+
+////////////////////////////////////////////////////////////////////////////////
+/// HASH TABLE CLASS
+////////////////////////////////////////////////////////////////////////////////
+
+class SequentialRobinHoodHashTable {
+private:
+  std::vector<SequentialBucket> buckets;
+  size_t length;
+  size_t size;
```

Comment:
Also I generally prefer to have private/protected member variables adhere to some kind of naming convention so they stick out (and also so I don't have to type `this` 😛). I'm partial to a prefix or postfix `_`.

Reply:
I think I'm going to stand up for using `this->var`. We can add a trailing `_` if you'd like, too. I'd be fine with `this->var_`.

Reply:
Bet. If we go with `this` we can drop the prefix/postfix since it'll always be clear.

Reply:
I just did both. That way the compiler enforces that the class variables are different too. Anything that makes it easier for a text editor (aka my brain at midnight) to find/replace